File: src/main/java/com/polygloat/model/Language.java
package com.polygloat.model;

import com.polygloat.dtos.request.LanguageDTO;
import lombok.Getter;
import lombok.Setter;
import lombok.ToString;

import javax.persistence.*;
import java.util.Set;

@Entity
@Table(uniqueConstraints = {
        @UniqueConstraint(columnNames = {"repository_id", "name"}, name = "language_repository_name"),
        @UniqueConstraint(columnNames = {"repository_id", "abbreviation"}, name = "language_abbreviation_name")
},
        indexes = {
                @Index(columnList = "abbreviation", name = "index_abbreviation"),
                @Index(columnList = "abbreviation, repository_id", name = "index_abbreviation_source")
        }
)
@ToString(of = {"id", "abbreviation", "name"})
public class Language extends AuditModel {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Getter
    @Setter
    private Long id;

    @OneToMany(fetch = FetchType.LAZY, mappedBy = "language")
    @Getter
    @Setter
    private Set<Translation> translations;

    @ManyToOne
    @Getter
    @Setter
    private Repository repository;

    @Getter
    @Setter
    private String abbreviation;

    @Getter
    @Setter
    private String name;

    public static Language fromRequestDTO(LanguageDTO dto) {
        Language language = new Language();
        language.setName(dto.getName());
        language.setAbbreviation(dto.getAbbreviation());
        return language;
    }

    public void updateByDTO(LanguageDTO dto) {
        this.name = dto.getName();
        this.abbreviation = dto.getAbbreviation();
    }
}

File: src/main/kotlin/io/tolgee/model/Language.kt
package io.tolgee.model

import io.tolgee.dtos.request.LanguageDto
import io.tolgee.service.dataImport.ImportService
import org.springframework.beans.factory.ObjectFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Configurable
import org.springframework.transaction.annotation.Transactional
import javax.persistence.*

@Entity
@EntityListeners(Language.Companion.LanguageListeners::class)
@Table(
        uniqueConstraints = [
            UniqueConstraint(
                    columnNames = ["project_id", "name"],
                    name = "language_project_name"
            ),
            UniqueConstraint(
                    columnNames = ["project_id", "tag"],
                    name = "language_tag_name")
        ],
        indexes = [
            Index(
                    columnList = "tag",
                    name = "index_tag"
            ),
            Index(
                    columnList = "tag, project_id",
                    name = "index_tag_project")
        ]
)
class Language : StandardAuditModel() {
    @OneToMany(fetch = FetchType.LAZY, mappedBy = "language")
    var translations: MutableSet<Translation>? = null

    @ManyToOne
    var project: Project? = null

    @Column(nullable = false)
    var tag: String? = null

    var name: String? = null

    @Column(nullable = false)
    var originalName: String? = null

    fun updateByDTO(dto: LanguageDto) {
        name = dto.name
        tag = dto.tag
    }

    override fun toString(): String {
        return "Language(tag=$tag, name=$name, originalName=$originalName)"
    }

    companion object {
        @JvmStatic
        fun fromRequestDTO(dto: LanguageDto): Language {
            val language = Language()
            language.name = dto.name
            language.tag = dto.tag
            language.originalName = dto.originalName
            return language
        }

        @Configurable
        class LanguageListeners {
            @Autowired
            lateinit var provider: ObjectFactory<ImportService>

            @PreRemove
            @Transactional
            fun preRemove(language: Language) {
                provider.`object`.onExistingLanguageRemoved(language)
            }
        }
    }
}
